lib/vector-core/src/event/finalization.rs

```diff
@@ -212,17 +236,18 @@ impl EventStatus {
     #[allow(clippy::match_same_arms)] // https://github.com/rust-lang/rust-clippy/issues/860
     pub fn update(self, status: Self) -> Self {
         match (self, status) {
-            // Recorded always overwrites existing status.
-            (_, Self::Recorded)
-            // Dropped always updates to the new status.
-                | (Self::Dropped, _) => status,
-            // Recorded is never updated.
-            (Self::Recorded, _)
-            // Delivered may update to `Failed`, but not to `Dropped`.
-                | (Self::Delivered, Self::Dropped)
-            // Failed does not otherwise update.
-                | (Self::Failed, _) => self,
-            (Self::Delivered, _) => status,
+            // `Recorded` always overwrites existing status and is never updated
+            (_, Self::Recorded) | (Self::Recorded, _) => Self::Recorded,
+            // `Dropped` always updates to the new status.
+            (Self::Dropped, _) => status,
+            // Updates *to* `Dropped` are nonsense.
+            (_, Self::Dropped) => self,
```

Comment:
If this is indicates a bug, should we panic? Or `debug_assert!(false)`.

Reply:
I would oppose a panic, as it is a harmless bug, but `debug_assert!` would make some sense to catch them in testing.

Reply:
Hah, except then I can't test that combination in the update matrix.